### PR TITLE
PUI Plugin Panels

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR serves as a "proof of concept" as to how client-side data could be served by plugins, in the new react / platform interface.

In this instance, custom "panels" are loaded via the API, and rendered on the page based on what is delivered to the server.

Currently this only serves raw text - much more work would be required to render:

- Raw html data
- Content rendered on server via template
- Content served via `/static/` and loaded dynamically
- Compiled react components etc

### Related Issues

- Closes https://github.com/inventree/InvenTree/issues/7709

### TODO

- [x] Update documentation
- [X] ~~include a way to embed javascript / typescript code into the documentation~~
- [X] Mark old plugin mixin as deprecated (in documentation at least)
- [x] Work out a way to support "themes" (observe dark mode, etc)
- [x] Test if passing context objects (such as user, navigate, api) works with "compiled" code (currently works with vite dev server)
- [x] Playwright unit tests

### Future Work

-  Make a "demo" plugin (published on PIP) which includes staticfiles, compiled javascript, etc

### Resources

Pinning some resources here for further reading:

- https://stackoverflow.com/questions/44778265/dynamically-loading-react-components
- https://stackoverflow.com/questions/50948995/react-application-with-external-plugins
- https://stackoverflow.com/questions/40249045/react-native-dynamically-load-modules-plugins
- https://samarjit-samanta.medium.com/react-dynamic-plugins-loading-e45905aa00a7
- https://webpack.js.org/configuration/target/
-

### CODE REVIEW

**Comment by wolflu05** [comment]
Thanks that you already started a draft for this long standing issue. I just want to add my thoughts about this now:
 
I See two ways how developers could build PUI integrations:

#### 1. They provide everything they need on thrown. E.g. they want to use react, they need to setup their own react spa, bundle it and ship the static files. Inventree PUI just loads embeds a specified entry html document.

| ✅ Advantages | ❌ Disadavantages |
|--------|--------|
| Plugin devs are not limited to any particular framework, they could use preact, react, solidjs, svelte, ... and even just plain HTML or vanilla js |  If multiple panel plugins using the same framework are installed and being serverd, multiple copies of the core frameworks code is needed to be loaded by the client |
| | Developers have to make their own efforts to align the UI of their plugin panel with the ui design of PUI, and rebuild helper functions, UI elements on their own |

To fix these disadvantages, I see two methods:
1. use a concept called module federation (search for the terms microfrontend, this concept is often used there). That way only one instance of react, mantine, ... needs to be used. But that would have the downside of version conflicts. E.g. if we do some breaking change in a component (e.g. add or remove a required prop), plugin panels using that federated component could just break 
2. ship the inventree ui and utility functions as npm package, panel developers could use. This would have the advantage, that developers have a predictable environment (e.g. they can pin this inventree ui dependency) and this works with all versions.

#### 2. They export a `render` function which gets called by PUI with a target DOM element they should render their content in. (Also with this method devs could use their framework of choice.) I Already made some tests with this a year ago: https://github.com/wolflu05/spa-preact-test

This would also add us some more possibilities, like we could let plugins provide some other UI things than just custom plugins. E.g. full new pages, custom label/report editors in the admin center (they also need a backchannel to talk to PUI).

---

Personally, I think I like a combination of all from above. We bundle all inventree components and utilities into a custom npm package developers could use. Yes, this adds the overhead of duplicate code, but I think that's much more reliable than somehow hack module federation to work correctly. Maybe we could module federation for the react dependency which has no breaking changes in minor version changes, and/or provide a global copy of preact, ... . Custom panels (and pages) use the HTML providing method (1), and custom editors could export a function (2).

**Comment by SchrodingersGat** [comment]
@wolflu05 some great ideas here. I think that whatever approach we choose:

- Enforce a single approach
- Provide some well-worked examples
- Publish a comprehensive set of common components
- Make it future-proof!

**Comment by SchrodingersGat** [comment]
@wolflu05 if you can provide a small POC for how we might load an entrypoint into a panel, that would be amazing.

**Comment by wolflu05** [comment]
The questions is what approach we should use. For my first suggestion, it would be as easy use using: [`dangerouslySetInnerHTML`](https://react.dev/reference/react-dom/components/common#dangerously-setting-the-inner-html)

And for the second, I actually made a POC last year [linked](https://github.com/wolflu05/spa-preact-test/blob/main/vite-spa/src/ExternalComponent.tsx) in my initial post. I also played a bit there with using the main react version + mantine to use in panels (commented out), but that doesn't worked well that way, as react was complaining, that the rendering react version is not the same instance as the one creating elements, but that might be an issue from me to that time.

**Comment by SchrodingersGat** [comment]
I would rather not use the first approach. I'll try to get your demo running when I have some time!

**Comment by matmair** [comment]
I really like @wolflu05 second idea; to keep the UI consistent providing a default set of UI components would be nice

**Comment by wolflu05** [comment]
Just one thought, maybe we can build the panel system in a way so that we later could also add those pages to the mobile app via a web view as an option (when plugin devs expose a panel, they could specify the target: web,mobile,...).

**Comment by matmair** [comment]
@SchrodingersGat regarding the ToDo section: why would we need javascript in the documentation?

**Comment by SchrodingersGat** [comment]
> @SchrodingersGat regarding the ToDo section: why would we need javascript in the documentation?

I want to be able to link in some source code from the frontend - namely to show what data gets passed through to the plugin context. And then to ensure that the docs stay up to date when we update / change that!

**Comment by wolflu05** [comment]
Regarding your is hidden function, can't we just return an object like
```json
{ hidden: true }
```
from the renderPanel function? We could then use this later to return some additional information too.
---

And just a question, how flexible will the render system be? I'm thinking about other places where we need to inject custom content, maybe we can build that system in a generic way, so that this can be later easily extended (currently I'm thinking about the following places):

### content based

1. custom panels on all pages using the panel group component (including admin center, system settings)
2. plugins settings drawer
3. custom editors/previews for templates
4. custom routes, e.g. to add completely new pages to the ui

### list based

5. custom table actions, row actions, line actions, ... that have an onclick handler where they could do stuff with the rows (also opening modals, drawers so they are half list, half content based)
6. include custom entries in the navigation

**Comment by matmair** [comment]
@SchrodingersGat that sounds like a good way to have the docs break / be un-reproducible every few weeks. Is this required or could we get by with a similar mechanism as the settings json to render the settings? HTML-only docs have proven to be very stable and I fear that adding another dependency to our data structures will make it even more difficult to maintain.

**Comment by SchrodingersGat** [comment]
@matmair to clarify, I do not want to run any javascript in the docs, just render code snippets (like we already do for our python source code)

**Comment by SchrodingersGat** [comment]
> And just a question, how flexible will the render system be? 

@wolflu05 I am designing this to be the first part of a flexible system, defined via `/api/plugin/ui/...`. I think it should be reasonably generic.

**Comment by SchrodingersGat** [comment]
Dynamic rendering from an imported javascript module is now working quite smoothly:

image

Still need to work out how to make them style nicely - i.e. build out a full working example that uses mantine and observes the correct theme.

File: src/frontend/src/hooks/UsePluginPanels.tsx [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        const iconName: string = props.icon || 'plugin';
+        const identifier = identifierString(
+          `plugin-panel-${props.plugin}-${props.name}`
+        );
+        const isHidden: boolean = panelState[identifier] ?? true;
+
+        return {
+          name: identifier,
+          label: props.label,
+          icon: <InvenTreeIcon icon={iconName as InvenTreeIconType} />,
```
Comment (wolflu05) [inline src/frontend/src/hooks/UsePluginPanels.tsx:115]: Q: Do you see a way here so we can use any tabler icon here? Not just the picked ones, available through the `InvenTreeIcon` component without adding much to the current bundle build size. Maybe we can use the APIIcon component here as a fallback?

**Review by wolflu05 — COMMENTED** [review]
All in all this LGTM. Thanks for stepping up to implement this. This is a huge step forward to the 1.0 milestone.

File: src/frontend/src/hooks/UsePluginPanels.tsx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  const globalSettings = useGlobalSettingsState();
+
+  const pluginPanelsEnabled: boolean = useMemo(
+    () => globalSettings.isSet('ENABLE_PLUGINS_INTERFACE'),
+    [globalSettings]
+  );
+
+  // API query to fetch initial information on available plugin panels
+  const { data: pluginData } = useQuery({
+    enabled: pluginPanelsEnabled && !!model && id !== undefined,
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/frontend/src/hooks/UsePluginPanels.tsx:45]: You can pass a string like "systemsettings" to the model, and `null` to the id.

File: src/frontend/src/components/nav/PanelGroup.tsx [hunk 3]
```diff
@@ -53,35 +58,39 @@ function BasePanelGroup({
...
+
+  // Hook to load plugins for this panel
+  const pluginPanels = usePluginPanels({
+    model: model,
+    instance: instance,
+    id: id
+  });
 
-  const setLastUsedPanel = useLocalState((state) =>
-    state.setLastUsedPanel(pageKey)
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/frontend/src/components/nav/PanelGroup.tsx:68]: Thanks for the report, fix this now

File: src/frontend/src/hooks/UsePluginPanels.tsx (hunk 1 above)
Comment (SchrodingersGat) [reply src/frontend/src/hooks/UsePluginPanels.tsx:132]: Yes, there is probably a way to re-use the existing icon framework! If you are ok with it, I'll leave that for a future work

File: src/backend/InvenTree/plugin/base/integration/UserInterfaceMixin.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        MIXIN_NAME = 'ui'
+
+    def __init__(self):
+        """Register mixin."""
+        super().__init__()
+        self.add_mixin('ui', True, __class__)
+
+    def get_custom_panels(
+        self, instance_type: str, instance_id: int, request: Request
+    ) -> list:
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply src/backend/InvenTree/plugin/base/integration/UserInterfaceMixin.py]: Nice idea, I have added this.

**Comment by matmair** [comment]
This looks very promising although some parts seem a bit brittle - is unit and e2e testing part of the scope for this PR?

**Comment by SchrodingersGat** [comment]
> is unit and e2e testing part of the scope for this PR?

Some testing will definitely be added! What parts in particular do you think look brittle?

**Comment by jacobfelknor** [comment]
@SchrodingersGat 

I apologize if this has been addressed elsewhere... but having trouble finding it if it is. I have a collection of CUI plugins (including panel plugins) that I've written and continue to maintain for my InvenTree instance. I use many "InvenTree builtins", like `inventreeTable()`, in my plugins in order to take advantage of all the reusability it has to offer. The switch to PUI, while exciting, makes me nervous that I'll have to rewrite all my plugins all at once. 

Is there a world where switching to PUI for the server and updating my plugins to also use PUI can be incremental for me? In other words, would it ever be possible for me to continue using my existing CUI plugins with PUI on the instance?

**Comment by SchrodingersGat** [comment]
@jacobfelknor this is a good point. I am going to write some documentation around panel plugin migration, before we force the switch.

**Comment by wolflu05** [comment]
I think first we need to publish the inventree npm package with the set of components, so that we can migrate e.g. the old form framework (I have used this in a panel) to the new react form framework. And same with other things.

**Pull request merged**